src/NuGet.Core/NuGet.Commands/PackCommand/Logging/PackageSpecificWarningProperties.cs

```diff
@@ -0,0 +1,121 @@
+    {
+        /// <summary>
+        /// Contains Package specific No warn properties.
+        /// NuGetLogCode -> LibraryId -> Set of Frameworks.
+        /// </summary>
+        public IDictionary<NuGetLogCode, IDictionary<string, ISet<NuGetFramework>>> Properties { get; private set; }
+
+        /// <summary>
+        /// Extracts PackageSpecific WarningProperties from a PackageSpec
+        /// </summary>
+        /// <param name="noWarnProperties">noWarnProperties containing the Dependencies with WarningProperties</param>
+        /// <returns>PackageSpecific WarningProperties extracted from a noWarnProperties</returns>
+        public static PackageSpecificWarningProperties CreatePackageSpecificWarningProperties(IDictionary<string, ISet<(NuGetLogCode, NuGetFramework)>> noWarnProperties)
+        {
+            if (noWarnProperties == null)
+                return null;
+
+            // NuGetLogCode -> LibraryId -> Set of Frameworks.
+            var warningProperties = new PackageSpecificWarningProperties();
+
+            foreach (KeyValuePair<string, ISet<(NuGetLogCode, NuGetFramework)>> packageNoWarnProperty in noWarnProperties)
+            {
+                string packageId = packageNoWarnProperty.Key;
+
+                foreach ((NuGetLogCode nuGetLogCode, NuGetFramework nuGetFramework) propertyPair in packageNoWarnProperty.Value)
+                {
+                    warningProperties.Add(propertyPair.nuGetLogCode, packageId, propertyPair.nuGetFramework);
+                }
+            }
+
+            return warningProperties;
+        }
+
+        /// <summary>
+        /// Adds a NuGetLogCode into the NoWarn Set for the specified library Id and target graph.
+        /// </summary>
+        /// <param name="code">NuGetLogCode for which no warning should be thrown.</param>
+        /// <param name="libraryId">Library for which no warning should be thrown.</param>
+        /// <param name="framework">Target graph for which no warning should be thrown.</param>
+        internal void Add(NuGetLogCode code, string libraryId, NuGetFramework framework)
+        {
+            if (Properties == null)
+            {
+                Properties = new Dictionary<NuGetLogCode, IDictionary<string, ISet<NuGetFramework>>>();
+            }
+
+            if (!Properties.ContainsKey(code))
+            {
+                Properties.Add(code, new Dictionary<string, ISet<NuGetFramework>>(StringComparer.OrdinalIgnoreCase));
+            }
+
+            if (Properties[code].ContainsKey(libraryId))
+            {
+                Properties[code][libraryId].Add(framework);
+            }
+            else
+            {
+                Properties[code].Add(libraryId, new HashSet<NuGetFramework>(new NuGetFrameworkFullComparer()) { framework });
+            }
+        }
+
+        public override int GetHashCode()
+        {
+            // return a constant hash for all objects since the contents of Properties are mutable
```

Comment:
I don't understand why the fact that a objec tis mutable means that we return a constant hash. 

Mutable objects can implement hashcode.

Reply:
It's existing logic I borrowed from here, it looks it's trying to emphasis it's not really meaningful, I don't believe we make use of hashcode of this object:

https://github.com/NuGet/NuGet.Client/blob/4f7ff0933e3d3a481b0ac74ef16a8e606350fb60/src/NuGet.Core/NuGet.Commands/RestoreCommand/Logging/PackageSpecificWarningProperties.cs#L155-L157

Reply:
That's extremely weird. 

Why are we even implementing it then?

Reply:
It looks since this class inherited from `IEquatable`, so it forces to implement GetHashCode(), but I don't why it have to inherit from `IEquatable`, I believe it's something to do with `TransitiveNoWarnUtils`. It's 4 years old code.

Reply:
Actually, I don't need it. I just adopted from restore logic, removed it.